numba/array_analysis.py

```diff
@@ -41,846 +52,1266 @@
-            c = self.array_shape_classes[arr_args[0].name][0]
-            for v in arr_args:
-                c = self._merge_classes(c, self.array_shape_classes[v.name][0])
-            new_class = self._get_next_class_with_size(len(arr_args))
-            return [c, new_class]
-        elif call_name in ['cumsum', 'cumprod']:
-            in_arr = args[0].name
-            in_ndims = self._get_ndims(in_arr)
-            # for 1D, output has same size
-            # TODO: return flattened size for multi-dimensional input
-            if in_ndims == 1:
-                return copy.copy(self.array_shape_classes[in_arr])
-        elif call_name == 'linspace':
-            # default is 50, arg3 is size
-            LINSPACE_DEFAULT_SIZE = 50
-            size = LINSPACE_DEFAULT_SIZE
-            if len(args) >= 3:
-                size = args[2].name
-            new_class = self._get_next_class_with_size(size)
-            return [new_class]
-        elif call_name == 'dot':
-            # https://docs.scipy.org/doc/numpy/reference/generated/numpy.dot.html
-            # for multi-dimensional arrays, last dimension of arg1 and second
-            # to last dimension of arg2 should be equal since used in dot product.
-            # if arg2 is 1D, its only dimension is used for dot product and
-            # should be equal to second to last of arg1.
-            assert len(args) == 2 or len(args) == 3
-            in1 = args[0].name
-            in2 = args[1].name
-            ndims1 = self._get_ndims(in1)
-            ndims2 = self._get_ndims(in2)
-            c1 = self.array_shape_classes[in1][ndims1 - 1]
-            c2 = UNKNOWN_CLASS
-
-            if ndims2 == 1:
-                c2 = self.array_shape_classes[in2][0]
+                    size = self._sum_size(equiv_set, [shape[i] for shape in shapes])
+                else:
+                    sizes = [ shape[i] for shape in shapes ]
+                    asserts.append(self._call_assert_equiv(scope, loc, equiv_set, sizes))
+                    size = sizes[0]
+                new_shape.append(size)
+        return tuple(new_shape), sum(asserts, [])
+
+    def _analyze_op_call_numpy_stack(self, scope, equiv_set, args, kws):
+        assert(len(args) > 0)
+        loc = args[0].loc
+        seq, op = find_build_sequence(self.func_ir, args[0])
+        n = len(seq)
+        require(n > 0)
+        axis = 0
+        if 'axis' in kws:
+            if isinstance(kws['axis'], int): # internal use only
+                axis = kws['axis']
             else:
-                c2 = self.array_shape_classes[in2][ndims2 - 2]
+                axis = find_const(self.func_ir, kws['axis'])
+        elif len(args) > 1:
+            axis = find_const(self.func_ir, args[1])
+        require(isinstance(axis, int))
+        # only build_tuple can give reliable count
+        require(op == 'build_tuple')
+        require(all(equiv_set.has_shape(x) for x in seq))
+        shapes = [equiv
```

Comment:
I'll remove it, since the generation of error message of `assert_equiv` was changed from runtime to compile-time, so it is no longer dependent on the variable names passed to `assert_equiv`.